File: frameworks/runtime-src/Classes/update/Localization.cpp
#include "Localization.h"
#include "cocos2d.h"
#include "external/json/document.h"
#include "2d/CCFontAtlasCache.h"

USING_NS_CC;
using namespace std;

Localization* Localization::_localization = NULL;
Localization::Localization()
{
}

Localization::~Localization()
{
}

Localization* Localization::getInstance()
{
	if (!_localization)
	{
		_localization = new Localization();
		_localization->loadText();
	}
	return _localization;
}


void Localization::loadText()
{

	localizedStrings.clear();

	string line, contents;

	// Get data of file
	contents = FileUtils::getInstance()->getStringFromFile("res/localize/vi");

	// Create a string stream so that we can use getline( ) on it
	istringstream fileStringStream(contents);

	// Get file contents line by line
	while (std::getline(fileStringStream, line))
	{
		if (line.find("/*", 0) == string::npos &&
			line.find("//", 0) == string::npos &&
			line.find("*/", 0) == string::npos) //filter the valid string of one line
		{
			std::string::size_type validPos = line.find('=', 0);
			if (validPos != string::npos)
			{
				std::string keyStr = line.substr(0, validPos - 1);
				// get valid string
				std::string subStr = line.substr(validPos + 1, line.size() - 1); 

				//trim space
				keyStr.erase(0, keyStr.find_first_not_of(" \t"));
				keyStr.erase(keyStr.find_last_not_of(" \t") + 1);

				subStr.erase(0, subStr.find_first_not_of(" \t"));
				subStr.erase(subStr.find_last_not_of(" \t") + 1);

				//trim \" 
				keyStr.erase(0, keyStr.find_first_not_of("\""));
				int findPosition = subStr.find('\"', 0);
				keyStr.erase(keyStr.find_last_not_of("\"") + 1);

				subStr.erase(0, subStr.find_first_not_of("\""));

				//trim ; character and last \" character
				subStr.erase(subStr.find_last_not_of(";") + 1);
				int position = subStr.find_last_not_of("\"") + 1;
				findPosition = subStr.find('\"', 0);
				if (findPosition != string::npos)
					subStr.erase(findPosition);

				//replace line feed with \n
				string::size_type pos(0);
				string old_value("\\n");
				if ((pos = subStr.find(old_value)) != string::npos)
				{
					for (; pos != string::npos; pos += 1)
					{
						if ((pos = subStr.find(old_value, pos)) != string::npos)
						{
							subStr.erase(pos, 2);
							subStr.insert(pos, 1, '\n');
						}
						else
							break;
					}
				}

				localizedStrings.insert(std::pair<std::string, std::string>(keyStr, subStr));
			}
		}
	}

}
std::string Localization::text(const char * mKey, const std::string& defaultText)
{
	std::map<std::string, std::string>::iterator itr = getInstance()->localizedStrings.find(std::string(mKey));
	if (itr != getInstance()->localizedStrings.end()) {
		return (itr->second).c_str();
	}
	if (!defaultText.empty())
	{
		return defaultText;
	}
	return mKey;
}

	

File: src/framework/Localization.js
/**
 * Created by KienVN on 5/27/2015.
 */
var g_localization = null;
fr.Localization = cc.Class.extend(
    {
        DEFAULT_DEVICE_LANGUAGE_CODE:"en",
        ctor:function()
        {
            this._dataConfig = this._loadConfig();
        },
        _loadConfig:function()
        {
            var fileName = "localize/config.json";
            var commonData = cc.loader.getRes(fileName);
            if (commonData == null) {
                cc.log("Load localization config error");
            }
            return commonData;
        },
        loadDefaultLanguage:function()
        {
            var defaultLangID = this._dataConfig.defaultLanguageCode;

            if (!defaultLangID)
            {
                defaultLangID = cc.sys.language;
                var isExistLang = false;
                for (var i in this._dataConfig.list)
                {
                    var data = this._dataConfig.list[i];
                    if (data.code == defaultLangID)
                    {
                        isExistLang = true;
                        break;
                    }
                }
                if (!isExistLang)
                {
                    defaultLangID = this.DEFAULT_DEVICE_LANGUAGE_CODE;
                }
            }
            this.setCurrentLanguage(defaultLangID);
        },
        setCurrentLanguage:function(code)
        {
            for (var i = 0; i < this._dataConfig.list.length; i++)
            {
                var data = this._dataConfig.list[i];
                if (data.code == code)
                {
                    if (this._currentLang != data)
                    {
                        this._currentLang = data;

                        this._loadTextForCurrentLanguage();
                    }
                    return;
                }
            }
            cc.log("ERROR: Localization::setCurrentLanguage: %s FALL!", code);
        },
        _loadTextForCurrentLanguage:function()
        {
            this._localizedStrings = {};
            var contents = null;
            cc.loader.loadTxt(this.getPathFileLang(this._currentLang.folder), function (error, txt) {
                if (error != null) {
                    cc.log("Load localization file error!");
                }
                else {
                    contents = txt;
                }
            });
            var lines = contents.split('\n');
            for(var i in lines)
            {
                var line = lines[i];
                if (line.indexOf("/*",0) == -1 &&
                    line.indexOf("//",0) == -1 &&
                    line.indexOf("*/",0) == -1) //filter the valid string of one line
                {
                    var validPos = line.indexOf('=', 0);
                    if (validPos != -1)
                    {
                        var keyStr = line.substring(0, validPos - 1);
                        // get valid string
                        var subStr = line.substring(validPos + 1, line.length - 1);

                        //trim space
                        keyStr = keyStr.slice(this.findFirstNotOf(keyStr," \t"));
                        keyStr = keyStr.slice(0,this.findLastNotOf(keyStr," \t") +1);

                        subStr = subStr.slice(this.findFirstNotOf(subStr," \t"));
                        subStr = subStr.slice(0,this.findLastNotOf(subStr," \t") +1);

                        //trim \"
                        keyStr = keyStr.slice(this.findFirstNotOf(keyStr,"\""));
                        keyStr = keyStr.slice(0,this.findLastNotOf(keyStr,"\"") +1);
                        var  findPosition = subStr.indexOf('\"', 0);
                        subStr = subStr.slice(this.findFirstNotOf(subStr,"\""));

                        //trim ; character and last \" character
                        subStr = subStr.slice(0,this.findLastNotOf(subStr,";") +1);
                        subStr = subStr.slice(0,this.findLastNotOf(subStr,"\"") +1);

                        //replace line feed with \n
                        subStr.replace(/\\n/g,"\n");

                        this._localizedStrings[keyStr] = subStr;
                    }
                }
            }

        },
        getText:function(key)
        {
            if(key in this._localizedStrings)
            {
                return this._localizedStrings[key];
            }
            cc.log("---- Localization text not found: "+ key );
            return key;
        },
        getCurrentLanguageName:function()
        {
            return this._currentLang.name;
        },
        getCurrentLanguageCode:function()
        {
            return this._currentLang.code;
        },
        getCountOfLanguages:function()
        {
            return this._dataConfig.list.length;
        },
        getLanguageNameAt:function(idx)
        {
            return this._dataConfig.list[idx].name;
        },
        getLanguageCodeAt:function(idx)
        {
            return this._dataConfig.list[idx].code;
        },
        getPathFileLang:function(code)
        {
            return "localize/" + code;
        },
        findLastNotOf:function(strSource,text) {
            var sourceLen = strSource.length;
            var strLen = text.length;
            if (strLen >sourceLen)
            {
                return -1;
            }
            var i = sourceLen - 1;
            while (i >= 0)
            {
                var result = false;
                for (var k = 0; k < strLen; k++)
                {
                    if (text[k] == strSource[i])
                    {
                        result = true;
                        break;
                    }
                }
                if(result)
                {
                    i-=1;
                }
                else
                {
                    return i;
                }
            }
            return -1;
        },
        findFirstNotOf:function(strSource, text) {
            var sourceLen = strSource.length;
            var strLen = text.length;
            var i = 0;
            while (i < sourceLen - 1) {
                var result = false;
                for (var k = 0; k < strLen; k++) {
                    if (text[k] == strSource[i]) {
                        result = true;
                        break;
                    }
                }
                if (result) {
                    i += 1;
                } else {
                    return i;
                }

            }
            return -1;
        }
    }
);
fr.Localization.getInstance = function(){
    if(g_localization == null)
    {
        g_localization = new fr.Localization();
        g_localization.loadDefaultLanguage();
    }
    return g_localization;
}
fr.Localization.text = function(key){
    return fr.Localization.getInstance().getText(key);
}
fr.Localization.ccsText = function(text)
{
    if(text.indexOf("lang_")==0)
        return fr.Localization.text(text);
    return text;
}